// uu_appg01_oidc-cmd/src/models/client-credentials-handler.js
"use strict";

const { generateKeyPair } = require('crypto');
const generateKeyPairAsync = generateKeyPair ? require("util").promisify(generateKeyPair) : null;
const NodeRsa = require("node-rsa");
const {Config} = require("uu_appg01_core-utils");
const {LoggerFactory} = require("uu_appg01_core-logging");
const {DaoFactory} = require("uu_appg01_datastore");
const {UseCaseContext} = require("uu_appg01_core-appserver");
const {SysAppConfigModel, SysAppWorkspaceConfigModel, SysKeyStoreModel} = require("uu_appg01_workspace");
const {ClientCredentialsProvider, Cache} = require("uu_appg01_oidc").Internal;

const ASID = Config.get("asid");
const KEY_SET = "uuAppOidcClientKey";
const DEFAULT_OIDCG01_PROVIDER_NAME = "oidcg01";
const UNREGISTERED_CLIENT_ID_PREFIX = "uu-oidc:unregistered-client:";
const LOGGER = LoggerFactory.get("uuapp.oidc.appserver.ClientCredentialsHandler");

/**
 * Component for managing of OidcClient credentials registration.
 */
class ClientCredentialsHandler {

  /**
   * Generates new private and public key to be used for OidcClient authentication.
   * @returns {{privateKey: *, publicKey: *}} Private and public key.
   */
  static async generateClientKeyPair() {
    if (generateKeyPairAsync) {
      return generateKeyPairAsync('rsa', {
        modulusLength: 2048,
        publicKeyEncoding: { type: 'spki', format: 'pem' },
        privateKeyEncoding: { type: 'pkcs8', format: 'pem' }
      });
    } else {
      let nodeRsa = new NodeRsa({b: 2048});
      nodeRsa.generateKeyPair();
      return {
        privateKey: nodeRsa.exportKey("private"),
        publicKey: nodeRsa.exportKey("public")
      };
    }
  }

  /**
   * Configures OidcClient for given AWID/ASID and service.
   * @param awid Awid or Asid
   * @param providerName Service name
   * @param clientId ClientId of registered OidcClient
   * @param keyPair Private and public key
   * @returns {Promise<null>} Void
   */
  static async configureOidcClient(awid, providerName, clientId, keyPair) {
    if (!DaoFactory.isDataStoreOn()) {
      return null;
    }
    if (awid === ASID) {
      let config = await SysAppConfigModel.getAppConfig(ASID);
      if (!config.oidcClientMap) {
        config.oidcClientMap = {};
      }
      config.oidcClientMap[providerName] = clientId;
      await SysAppConfigModel.setAppConfig(ASID, config);
    } else {
      let config = await SysAppWorkspaceConfigModel.getConfiguration(awid);
      if (!config.oidcClientMap) {
        config.oidcClientMap = {};
      }
      config.oidcClientMap[providerName] = clientId;
      await SysAppWorkspaceConfigModel.setConfiguration(awid, config);
    }
    await SysKeyStoreModel.putKey(ASID, KEY_SET, `${providerName}|${clientId}|Private`, keyPair.privateKey);
    await SysKeyStoreModel.putKey(ASID, KEY_SET, `${providerName}|${clientId}|Public`, keyPair.publicKey);
    return null;
  }

  /**
   * Returns clientId registered for given AWID/ASID and service.
   * @param awid Awid or Asid
   * @param providerName Service name
   * @returns {Promise<null|*>} Registered clientId
   */
  static async getClientId(awid, providerName) {
    if (!DaoFactory.isDataStoreOn() || providerName === DEFAULT_OIDCG01_PROVIDER_NAME) {
      return null;
    }
    let config;
    if (awid === ASID) {
      config = await SysAppConfigModel.getAppConfig(ASID);
    } else {
      config = await SysAppWorkspaceConfigModel.getConfiguration(awid);
    }
    if (config && config.oidcClientMap && config.oidcClientMap[providerName]) {
      return config.oidcClientMap[providerName];
    } else {
      return null;
    }
  }

  /**
   * Returns client credentials based on AWID/ASID from current
   * use case context and given service.
   * @param providerName Name of service
   * @returns {Promise<{clientId, clientPrivateKey}|null>} Client credentials
   * @see ClientCredentialsProvider.getClientCredentials
   */
  static async getClientCredentials(providerName) {
    if (!DaoFactory.isDataStoreOn() || providerName === DEFAULT_OIDCG01_PROVIDER_NAME) {
      return null;
    }
    let awid = this._getAwidFromContext();
    if (!awid) {
      LOGGER.debug("Unable to get awid from use case context (probably due to invocation outside of use case scope). Using OidcClient from asid.");
      return await this._getCredentialsFromAsid(providerName);
    }
    if (awid === ASID) {
      return await this._getCredentialsFromAsid(providerName);
    } else {
      return (await this._getCredentialsFromAwid(awid, providerName)) || (await this._getCredentialsFromAsid(providerName));
    }
  }

  /**
   * Returns credentials of unregistered client composed of AWID/ASID from
   * current use case context.
   * @param providerName Name of servive
   * @returns {Promise<{clientId: string, clientSecret: string}|null>} Unregistered client credentials
   * @see ClientCredentialsProvider.getClientCredentials
   */
  static async getUnregisteredCredentials(providerName) {
    let awid = this._getAwidFromContext();
    if (!awid) {
      return null;
    }
    let clientId = `${UNREGISTERED_CLIENT_ID_PREFIX}${awid}`;
    let clientSecret = "unregistered";
    return {clientId, clientSecret};
  }

  static _getAwidFromContext() {
    if (!UseCaseContext.isAvailable()) {
      return null;
    }
    let uri = UseCaseContext.getUri();
    return uri ? uri.getAwid() : null;
  }

  static async _getCredentialsFromAwid(awid, providerName) {
    let config;
    try {
      config = await SysAppWorkspaceConfigModel.getConfiguration(awid);
    } catch (e) {
      LOGGER.error(`Unable to load OidcClient for awid ${awid} and service ${providerName}. Using OidcClient from asid.`, e);
      return null;
    }
    if (!config || !config.oidcClientMap || !config.oidcClientMap[providerName]) {
      LOGGER.warn(`Missing configuration of OidcClient for awid ${awid} and service ${providerName}. Using OidcClient from asid.`);
      return null;
    }
    let clientId = config.oidcClientMap[providerName];
    let clientPrivateKey;
    try {
      clientPrivateKey = await SysKeyStoreModel.getKey(ASID, KEY_SET, `${providerName}|${clientId}|Private`);
    } catch (e) {
      LOGGER.error(`Unable to load credentials of OidcClient ${clientId} for awid ${awid}. Using OidcClient from asid.`, e);
      return null;
    }
    if (!clientPrivateKey) {
      LOGGER.warn(`Missing configuration of OidcClient for awid ${awid} and service ${providerName}. Using OidcClient from asid.`);
      return null;
    }
    return {clientId, clientPrivateKey: clientPrivateKey.key};
  }

  static async _getCredentialsFromAsid(providerName) {
    let config;
    try {
      config = await SysAppConfigModel.getAppConfig(ASID);
    } catch (e) {
      LOGGER.error(`Unable to load OidcClient for asid ${ASID} and service ${providerName}. Using default OidcClient from configuration.`, e);
      return null;
    }
    if (!config || !config.oidcClientMap || !config.oidcClientMap[providerName]) {
      LOGGER.warn(`Missing configuration of OidcClient for asid ${ASID} and service ${providerName}. Using default OidcClient from configuration.`);
      return null;
    }
    let clientId = config.oidcClientMap[providerName];
    let clientPrivateKey;
    try {
      clientPrivateKey = await SysKeyStoreModel.getKey(ASID, KEY_SET, `${providerName}|${clientId}|Private`);
    } catch (e) {
      LOGGER.error(`Unable to load credentials of OidcClient ${clientId} for asid ${ASID}. Using default OidcClient from configuration.`, e);
      return null;
    }
    if (!clientPrivateKey) {
      LOGGER.warn(`Missing configuration of OidcClient for asid ${ASID} and service ${providerName}. Using default OidcClient from configuration.`);
      return null;
    }
    return {clientId, clientPrivateKey: clientPrivateKey.key};
  }

}

// Register component as custom credentials provider
ClientCredentialsProvider.setExternalProvider(ClientCredentialsHandler);

module.exports = ClientCredentialsHandler;
