// uu_appg01_workspace/src/errors/sys-app-workspace-config-errors.js
const WorkspaceError = require("./workspace-error.js");

const SysGetAppWorkspaceConfig = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysGetAppWorkspaceConfig/`,

  ConfigDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Object sysAppWorkspaceConfig does not exist.";
      this.code = `${SysGetAppWorkspaceConfig.UC_CODE}sysAppWorkspaceConfigDoesNotExist`;
    }
  }
};

const SysSetAppWorkspaceConfig = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysSetAppWorkspaceConfig/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysSetAppWorkspaceConfig.UC_CODE}invalidDtoIn`;
    }
  },

  DaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update sysAppWorkspaceConfig by sysAppWorkspaceConfig DAO update failed.";
      this.code = `${SysSetAppWorkspaceConfig.UC_CODE}sysAppWorkspaceConfigDaoUpdateFailed`;
    }
  },

  MaxDtoInSizeExceeded: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "The size of the shape must not exceed 16000 B.";
      this.code = `${SysSetAppWorkspaceConfig.UC_CODE}MaxDtoInSizeExceeded`;
    }
  }
};

const SysClearAppWorkspaceConfig = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysClearAppWorkspaceConfig/`,

  ConfigDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "uuObject sysAppWorkspaceConfig does not exist.";
      this.code = `${SysClearAppWorkspaceConfig.UC_CODE}sysAppWorkspaceConfigDoesNotExist`;
    }
  },

  DaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update sysAppWorkspaceConfig by sysAppWorkspaceConfig Dao update failed.";
      this.code = `${SysClearAppWorkspaceConfig.UC_CODE}sysAppWorkspaceConfigDaoUpdateFailed`;
    }
  }
};

module.exports = {
  SysSetAppWorkspaceConfig,
  SysGetAppWorkspaceConfig,
  SysClearAppWorkspaceConfig
};
