// uu_appg01_workspace/src/errors/sys-app-config-errors.js
const WorkspaceError = require("./workspace-error.js");

const SysInitApp = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysInitApp/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysInitApp.UC_CODE}invalidDtoIn`;
    }
  },
  AsidAlreadyInitializedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Asid has been already initialized.";
      this.code = `${SysInitApp.UC_CODE}asidAlreadyInitialized`;
    }
  },
  SysAppConfigDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of sysAppConfig by Dao create failed.";
      this.code = `${SysInitApp.UC_CODE}sysAppConfigDaoCreateFailed`;
    }
  }
};

const SysSetAppConfig = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysSetAppConfig/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysSetAppConfig.UC_CODE}invalidDtoIn`;
    }
  },
  SysAppConfigDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update sysAppConfig by sysAppConfig Dao update failed.";
      this.code = `${SysSetAppConfig.UC_CODE}sysAppConfigDaoUpdateFailed`;
    }
  }
};

const SysGetAppConfig = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysGetAppConfig/`,

  SysAppConfigDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Object sysAppConfig does not exist.";
      this.code = `${SysGetAppConfig.UC_CODE}sysAppConfigDoesNotExist`;
    }
  }
};

module.exports = {
  SysInitApp,
  SysSetAppConfig,
  SysGetAppConfig
};
