// uu_appg01_workspace/src/errors/sys-app-workspace-errors.js
const WorkspaceError = require("./workspace-error.js");
const { BaseError } = require("uu_appg01_core-utils");

const SysInitAppWorkspace = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysInitAppWorkspace/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysInitAppWorkspace.UC_CODE}invalidDtoIn`;
    }
  },

  AwidNotUniqueError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "sysAppWorkspace already initialized.";
      this.code = `${SysInitAppWorkspace.UC_CODE}awidNotUnique`;
    }
  },

  AwidConflictsWithAsidError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "sysAppWorkspace cannot be initialized.";
      this.code = `${SysInitAppWorkspace.UC_CODE}awidConflictsWithAsid`;
    }
  },

  CreateSchemaFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of uuObject schema by uuObject DAO createSchema failed.";
      this.code = `${SysInitAppWorkspace.UC_CODE}uuObjectDaoCreateSchemaFailed`;
    }
  },

  SysAppWorkspaceDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of sysAppWorkspace by sysAppWorkspace DAO create failed.";
      this.code = `${SysInitAppWorkspace.UC_CODE}sysAppWorkspaceDaoCreateFailed`;
    }
  },

  SysProfileDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of sysProfile by DAO create failed.";
      this.code = `${SysInitAppWorkspace.UC_CODE}sysProfileDaoCreateFailed`;
    }
  },

  SysAppWorkspaceConfigDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of sysAppWorkspaceConfig by DAO create failed.";
      this.code = `${SysInitAppWorkspace.UC_CODE}sysAppWorkspaceConfigDaoCreateFailed`;
    }
  },

  SysProfileDaoDeleteByAwidFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Delete of sysProfile by sysProfile Dao deleteByAwid failed.";
      this.code = `${SysInitAppWorkspace.UC_CODE}sysProfileDaoDeleteByAwidFailed`;
    }
  },

  SysAppWorkspaceConfigDaoDeleteByAwidFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Delete of sysAppWorkspaceConfig by sysAppWorkspaceConfig Dao delete failed.";
      this.code = `${SysInitAppWorkspace.UC_CODE}sysAppWorkspaceConfigDaoDeleteByAwidFailed`;
    }
  },

  SysAppWorkspaceDaoDeleteByAwidFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Delete of sysAppWorkspace by sysAppWorkspace Dao delete failed.";
      this.code = `${SysInitAppWorkspace.UC_CODE}sysAppWorkspaceDaoDeleteByAwidFailed`;
    }
  },

  SysAppWorkspaceDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update of sysAppWorkspace by sysAppWorkspace Dao updateByAwid failed.";
      this.code = `${SysInitAppWorkspace.UC_CODE}sysAppWorkspaceDaoUpdateFailed`;
    }
  },

  SysAppWorkspaceDataStoreStatsDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Creation of sysDataStoreStats by DAO create failed.";
      this.code = `${SysInitAppWorkspace.UC_CODE}sysAppWorkspaceDataStoreStatsDaoCreateFailedError`;
    }
  }
};

const SysGetAppWorkspace = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysGetAppWorkspace/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysGetAppWorkspace.UC_CODE}invalidDtoIn`;
    }
  },

  SysAppWorkspaceDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Object sysAppWorkspace does not exist.";
      this.code = `${SysGetAppWorkspace.UC_CODE}sysAppWorkspaceDoesNotExist`;
    }
  }
};

const SysGetLicenseOwner = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysGetLicenseOwner/`,

  SysAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Object sysAppWorkspace does not exist.";
      this.code = `${SysGetLicenseOwner.UC_CODE}sysAppWorkspaceDoesNotExist`;
    }
  }
};

const SetAuthorizationType = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysSetAuthType/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SetAuthorizationType.UC_CODE}invalidDtoIn`;
    }
  },
  UuAppWorkspaceIsNotInProperStateError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "sysAppWorkspace is not in proper state.";
      this.code = `${SetAuthorizationType.UC_CODE}uuAppWorkspaceIsNotInProperState`;
    }
  },
  UserIsNotAuthorizedToChangeAuthStrategy: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Logged user is not authorized to change authorization strategy.";
      this.code = `${SetAuthorizationType.UC_CODE}userIsNotAuthorizedToChangeAuthStrategy`;
    }
  },
  SysAppWorkspaceDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update of sysAppWorkspace by sysAppWorkspace Dao updateByAwid failed.";
      this.code = `${SetAuthorizationType.UC_CODE}sysAppWorkspaceDaoUpdateFailed`;
    }
  }
};

const ConnectArtifact = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysConnectArtifact/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${ConnectArtifact.UC_CODE}invalidDtoIn`;
    }
  },
  UuAppWorkspaceIsNotInProperStateError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "sysAppWorkspace is not in proper state.";
      this.code = `${ConnectArtifact.UC_CODE}uuAppWorkspaceIsNotInProperState`;
    }
  },
  UuCmdConnectAppWorkspaceCallFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "uuCMD uuAwsc/connectAppWorkspace could not be started.";
      this.code = `${ConnectArtifact.UC_CODE}uuCmdConnectAppWorkspaceCallFailed`;
    }
  },
  SysAppWorkspaceDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update of sysAppWorkspace by sysAppWorkspace Dao updateByAwid failed.";
      this.code = `${ConnectArtifact.UC_CODE}sysAppWorkspaceDaoUpdateFailed`;
    }
  },
  TokenCreateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message =
        "Failed to create uuAppClientToken because keys for it's signature were not initialized. Run the uuCmd sys/appClientToken/initKeys first.";
      this.code = `${ConnectArtifact.UC_CODE}uuAppClientTokenCreateFailed`;
    }
  }
};

const SysSetAppWorkspaceState = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysSetAppWorkspaceState/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysSetAppWorkspaceState.UC_CODE}invalidDtoIn`;
    }
  },

  SysAppWorkspaceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update of sysAppWorkspace by sysAppWorkspace Dao updateByAwid failed.";
      this.code = `${SysSetAppWorkspaceState.UC_CODE}sysAppWorkspaceDaoUpdateFailed`;
    }
  },

  SysAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysAppWorkspace does not exist.";
      this.code = `${SysSetAppWorkspaceState.UC_CODE}sysAppWorkspaceDoesNotExist`;
    }
  }
};

const SysDeleteAppWorkspace = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysDeleteAppWorkspace/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysDeleteAppWorkspace.UC_CODE}invalidDtoIn`;
    }
  },
  AwidNotClosed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysAppWorkspace must be in the closed state to perform a delete operation.";
      this.code = `${SysDeleteAppWorkspace.UC_CODE}sysAppWorkspaceIsNotClosed`;
    }
  },
  SysAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysAppWorkspace does not exist.";
      this.code = `${SysDeleteAppWorkspace.UC_CODE}sysAppWorkspaceDoesNotExist`;
    }
  },
  ArtifactConnectedToSysAppWorkspace: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "It is not possible to delete uuAppWorkspace to which is connected an artifact.";
      this.code = `${SysDeleteAppWorkspace.UC_CODE}artifactConnectedToSysAppWorkspace`;
    }
  }
};

const SysSetAwidRuntimeMode = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysSetAwidRuntimeMode/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysSetAwidRuntimeMode.UC_CODE}invalidDtoIn`;
    }
  },
  SysAppWorkspaceDaoUpdateFailed: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update runtime mode by sysAppWorkspace DAO update failed.";
      this.code = `${SysSetAwidRuntimeMode.UC_CODE}sysAppWorkspaceDaoUpdateFailed`;
    }
  }
};

const SysGetAwidRuntimeMode = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysGetAwidRuntimeMode/`,

  SysAppWorkspaceDoesNotExist: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Object sysAppWorkspace does not exist.";
      this.code = `${SysGetAwidRuntimeMode.UC_CODE}sysAppWorkspaceDoesNotExist`;
    }
  }
};

class MaximumNumberAwidsError extends BaseError {
  constructor(maxNoa) {
    super(`Inserting uuObject failed. Maximum number of awids (${maxNoa}) reached.`);
  }
}

module.exports = {
  SysInitAppWorkspace,
  SysGetAppWorkspace,
  MaximumNumberAwidsError,
  SetAuthorizationType,
  ConnectArtifact,
  SysSetAppWorkspaceState,
  SysDeleteAppWorkspace,
  SysSetAwidRuntimeMode,
  SysGetAwidRuntimeMode,
  SysGetLicenseOwner
};
