// uu_appg01_workspace/src/errors/sys-profiles-errors.js
const WorkspaceError = require("./workspace-error.js");

const SysSetProfile = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysSetProfile/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysSetProfile.UC_CODE}invalidDtoIn`;
    }
  },

  DaoUpdateByCodeFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update sysProfile by sysProfile DAO updateByCode failed.";
      this.code = `${SysSetProfile.UC_CODE}sysProfileDaoUpdateByCodeFailed`;
    }
  }
};

const SysGetProfile = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysGetProfile/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysGetProfile.UC_CODE}invalidDtoIn`;
    }
  }
};

const SysListProfiles = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysListProfiles/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysListProfiles.UC_CODE}invalidDtoIn`;
    }
  },

  DaoListByUuIdAndProfileError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "List sysProfile by sysProfile DAO listByUuIdAndProfile failed.";
      this.code = `${SysListProfiles.UC_CODE}sysProfileDaoGetByCodeFailed`;
    }
  }
};

const SysUpdateProfiles = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysUpdateProfiles/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysUpdateProfiles.UC_CODE}invalidDtoIn`;
    }
  },
  SysAppWorkspaceDoesNotExistError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "SysAppWorkspace does not exist or is not in initialized state.";
      this.code = `${SysUpdateProfiles.UC_CODE}sysAppWorkspaceDoesNotExist`;
    }
  },
  SysAppWorkspaceDaoUpdateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Update of sysAppWorkspace by sysAppWorkspace Dao updateByAwid failed.";
      this.code = `${SysUpdateProfiles.UC_CODE}sysAppWorkspaceDaoUpdateFailed`;
    }
  },
  AppNotInPrivilegedModeError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "The application is not in the privileged runtime mode that is required for this uuCmd.";
      this.code = `${SysUpdateProfiles.UC_CODE}appNotInPrivilegedMode`;
    }
  },
  SysProfileDaoCreateFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Create of sysProfile by Dao create failed.";
      this.code = `${SysUpdateProfiles.UC_CODE}sysProfileDaoCreateFailed`;
    }
  },
  SysProfileDaoDeleteByCodeFailedError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "Delete of sysProfile by Dao deleteByCode failed.";
      this.code = `${SysUpdateProfiles.UC_CODE}sysProfileDaoDeleteByCodeFailed`;
    }
  }
};

module.exports = {
  SysSetProfile,
  SysGetProfile,
  SysListProfiles,
  SysUpdateProfiles
};
