// uu_appg01_workspace/src/errors/sys-data-store-errors.js
const WorkspaceError = require("./workspace-error.js");

const SysDropDataStore = {
  UC_CODE: `${WorkspaceError.ERROR_PREFIX}sysDropDataStores/`,

  InvalidDtoInError: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "DtoIn is not valid.";
      this.code = `${SysDropDataStore.UC_CODE}invalidDtoIn`;
    }
  },
  AppNotInPrivilegedMode: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "The application is not in the privileged runtime mode that is required for this uuCmd.";
      this.code = `${SysDropDataStore.UC_CODE}appNotInPrivilegedMode`;
    }
  },
  DevelopmentModeNotActive: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "The application is not deployed with development mode enabled.";
      this.code = `${SysDropDataStore.UC_CODE}developmentModeNotActive`;
      this.status = 503;
    }
  },
  DataStoreDisabled: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message =
        "The application has disabled the persistence layer. Thus, it does not have any uuSubAppDataStore to drop.";
      this.code = `${SysDropDataStore.UC_CODE}dataStoreDisabled`;
    }
  },
  DataStoreNotConfigured: class extends WorkspaceError {
    constructor() {
      super(...arguments);
      this.message = "The uuSubAppDataStore specified in dtoIn is not configured in the application.";
      this.code = `${SysDropDataStore.UC_CODE}dataStoreNotConfigured`;
    }
  }
};

module.exports = {
  SysDropDataStore
};
