// uu_appg01_auditlog/src/dao/sys-audit-log-record-mongo-db.js
"use strict";

const {UuObjectDao} = require("uu_appg01_objectstore");

class SysAuditLogRecordMongoDB extends UuObjectDao {
  async createSchema() {
    await super.createIndex({awid: 1, _id: 1}, {unique: true});
    await super.createIndex({awid: 1, logTime: -1, spp: 1});
    await super.createIndex({awid: 1, logTime: -1, identity: 1});
  }

  async create(uuObject) {
    return await super.insertOne(uuObject);
  }

  async listByTimeAndSpp(awid, timeFrom, timeTo, spp, pageInfo = {}) {
    let filter = {
      awid,
      logTime: {
        "$lte": timeTo
      }
    };

    if (timeFrom) {
      filter.logTime["$gte"] = timeFrom;
    }

    if (typeof spp === 'string' || spp instanceof String) {
      filter.spp = spp;
    }

    return await super.find(filter, pageInfo);
  }

  async listByTimeAndIdentity(awid, timeFrom, timeTo, identity, pageInfo) {
    let filter = {
      awid,
      logTime: {
        "$lte": timeTo
      }
    };

    if (timeFrom) {
      filter.logTime["$gte"] = timeFrom;
    }

    if (identity) {
      filter.identity = identity;
    }

    return await super.find(filter, pageInfo);
  }

  async deleteBeforeTime(awid, toTime) {
    let filter = {
      awid,
      logTime: {
        "$lte": toTime
      }
    };

    return await super.deleteMany(filter);
  }
}

module.exports = SysAuditLogRecordMongoDB;
