// uu_appg01_auditlog/src/api/errors/sys-auditlog-error.js
"use strict";

const SysAuditLogGeneralError = require("./sys-auditlog-general-error.js");

const CreateAuditLogRecord = {
  UC_CODE: `${SysAuditLogGeneralError.ERROR_PREFIX}createAuditLogRecord/`,

  InvalidDtoIn: class extends SysAuditLogGeneralError {
    constructor() {
      super(...arguments);
      this.code = `${CreateAuditLogRecord.UC_CODE}invalidDtoIn`;
      this.message = "DtoIn is not valid.";
    }
  },

  DaoFailed: class extends SysAuditLogGeneralError {
    constructor() {
      super(...arguments);
      this.code = `${CreateAuditLogRecord.UC_CODE}daoFailed`;
      this.message = "DAO method failed.";
    }
  }
};

const ListByTimeAndSpp = {
  UC_CODE: `${SysAuditLogGeneralError.ERROR_PREFIX}listByTimeAndSpp/`,

  InvalidDtoIn: class extends SysAuditLogGeneralError {
    constructor() {
      super(...arguments);
      this.code = `${ListByTimeAndSpp.UC_CODE}invalidDtoIn`;
      this.message = "DtoIn is not valid.";
    }
  },

  InvalidIntervalDefinition: class extends SysAuditLogGeneralError {
    constructor() {
      super(...arguments);
      this.code = `${ListByTimeAndSpp.UC_CODE}invalidIntervalDefinition`;
      this.message = "DtoIn.timeFrom is not before or equal to dtoIn.timeTo.";
    }
  },

  DaoFailed: class extends SysAuditLogGeneralError {
    constructor() {
      super(...arguments);
      this.code = `${ListByTimeAndSpp.UC_CODE}daoFailed`;
      this.message = "DAO method failed.";
    }
  }
};

const ListMyByTime = {
  UC_CODE: `${SysAuditLogGeneralError.ERROR_PREFIX}listMyByTime/`,

  InvalidDtoIn: class extends SysAuditLogGeneralError {
    constructor() {
      super(...arguments);
      this.code = `${ListMyByTime.UC_CODE}invalidDtoIn`;
      this.message = "DtoIn is not valid.";
    }
  },

  InvalidIntervalDefinition: class extends SysAuditLogGeneralError {
    constructor() {
      super(...arguments);
      this.code = `${ListMyByTime.UC_CODE}invalidIntervalDefinition`;
      this.message = "DtoIn.timeFrom is not before or equal to dtoIn.timeTo.";
    }
  },

  DaoFailed: class extends SysAuditLogGeneralError {
    constructor() {
      super(...arguments);
      this.code = `${ListMyByTime.UC_CODE}daoFailed`;
      this.message = "DAO method failed.";
    }
  }
};

const DeleteBeforeTime = {
  UC_CODE: `${SysAuditLogGeneralError.ERROR_PREFIX}deleteBeforeTime/`,

  InvalidDtoIn: class extends SysAuditLogGeneralError {
    constructor() {
      super(...arguments);
      this.code = `${DeleteBeforeTime.UC_CODE}invalidDtoIn`;
      this.message = "DtoIn is not valid.";
    }
  },

  DaoFailed: class extends SysAuditLogGeneralError {
    constructor() {
      super(...arguments);
      this.code = `${DeleteBeforeTime.UC_CODE}daoFailed`;
      this.message = "DAO method failed.";
    }
  }
};

const AppLogStoreImport = {
  UC_CODE: `${SysAuditLogGeneralError.ERROR_PREFIX}importToAppLogStore/`,

  ImportFailed: class extends SysAuditLogGeneralError {
    constructor() {
      super(...arguments);
      this.code = `${AppLogStoreImport.UC_CODE}importFailed`;
      this.message = "Failed to import logs to the uuAppLogStore.";
    }
  }
};

module.exports = {
  CreateAuditLogRecord,
  ListByTimeAndSpp,
  ListMyByTime,
  DeleteBeforeTime,
  AppLogStoreImport
};
