// uu_appg01_workspace/index.js
"use strict";

module.exports = {
  get WorkspaceAuthorizationService() {
    return require("./src/workspace-authorization-service.js");
  },
  get WorkspaceAuthorizationResult() {
    return require("./src/workspace-authorization-result.js");
  },
  get WorkspaceError() {
    return require("./src/errors/workspace-error.js");
  },
  get AppClientTokenService() {
    return require("./src/app-client-token-service.js");
  },
  get AppClientToken() {
    return require("./src/app-client-token.js");
  },
  get AppKey() {
    return require("./src/app-key.js");
  },
  get ProfileConfig() {
    return require("./src/helpers/profile-config.js");
  },

  // TODO We need some workspace API
  // (Models should not be accessed directly)
  get SysAppWorkspaceConfigModel() {
    return require("./src/models/sys-app-workspace-config-model.js");
  },
  get SysAppWorkspaceConfigErrors() {
    return require("./src/errors/sys-app-workspace-config-errors.js");
  },
  get SysAppWorkspaceModel() {
    return require("./src/models/sys-app-workspace-model.js");
  },
  get SysAppWorkspaceErrors() {
    return require("./src/errors/sys-app-workspace-errors.js");
  },
  get SysPermissionModel() {
    return require("./src/models/sys-permission-model.js");
  },
  get SysPermissionErrors() {
    return require("./src/errors/sys-permission-errors.js");
  },
  get SysProfileModel() {
    return require("./src/models/sys-profile-model.js");
  },
  get SysProfileErrors() {
    return require("./src/errors/sys-profiles-errors.js");
  },
  get SysDataStoreStatsModel() {
    return require("./src/models/sys-data-store-stats-model.js");
  },
  get SysDataStoreStatsErrors() {
    return require("./src/errors/sys-data-store-stats-errors.js");
  },
  get SysAppConfigModel() {
    return require("./src/models/sys-app-config-model.js");
  },
  get SysAppConfigErrors() {
    return require("./src/errors/sys-app-config-errors.js");
  },
  get SysAppRuntimeModeModel() {
    return require("./src/models/sys-app-runtime-mode-model.js");
  },
  get SysAppRuntimeModeErrors() {
    return require("./src/errors/sys-app-runtime-mode-errors.js");
  },
  get SysDumpRestoreModel() {
    return require("./src/models/sys-dump-restore-model.js");
  },
  get SysDumpRestoreErrors() {
    return require("./src/errors/sys-dump-restore-errors.js");
  },
  get SysAppClientTokenModel() {
    return require("./src/models/sys-app-client-token-model.js");
  },
  get SysAppClientTokenErrors() {
    return require("./src/errors/sys-app-client-token-errors.js");
  },
  get SysKeyStoreModel() {
    return require("./src/models/sys-key-store-model.js");
  },
  get SysKeyStoreErrors() {
    return require("./src/errors/sys-key-store-errors.js");
  }
};
