// uu_appg01_workspace/src/daos/sys-app-workspace-mongo.js
"use strict";

const { UuObjectDao } = require("uu_appg01_objectstore");

class SysAppWorkspaceMongo extends UuObjectDao {
  async createSchema() {
    await super.createIndex(
      {
        awid: 1
      },
      { unique: true }
    );
  }

  create(appWorkspace) {
    return super.insertOne(appWorkspace);
  }

  getByAwid(awid) {
    return super.findOne({ awid: awid });
  }

  getByArtifactUri(artUri) {
    return super.findOne({ artifactUri: artUri });
  }

  updateByAwid(awid, appWorkspace) {
    return super.findOneAndUpdate({ awid }, appWorkspace, "NONE");
  }

  updateState(awid, state) {
    return super.findOneAndUpdate({ awid }, { state }, "NONE");
  }

  updateAwidRuntimeMode(awid, runtimeMode) {
    return super.findOneAndUpdate({ awid }, { runtimeMode }, "NONE");
  }

  update(appWorkspace) {
    return super.findOneAndUpdate(
      {
        id: appWorkspace["id"],
        awid: appWorkspace["awid"]
      },
      appWorkspace,
      "NONE"
    );
  }

  deleteByAwid(awid) {
    return super.deleteOne({ awid: awid });
  }

  getCount() {
    return super.count({});
  }

  listByState(state, pageInfo = {}) {
    return super.find({ state }, pageInfo);
  }

  listByStateAndAwidList(state, awidList, pageInfo = {}) {
    let filter = {
      state,
      awid: { $in: awidList }
    };
    return super.find(filter, pageInfo);
  }
}

module.exports = SysAppWorkspaceMongo;
