// uu_appg01_workspace/src/daos/sys-profile-mongo.js
"use strict";

const { UuObjectDao } = require("uu_appg01_objectstore");

class SysProfileMongo extends UuObjectDao {
  async createSchema() {
    await super.createIndex({ awid: 1, code: 1 }, { unique: true });
  }

  create(input) {
    if (Array.isArray(input)) {
      return super.insertMany(input);
    } else {
      return super.insertOne(input);
    }
  }

  get(awid) {
    return super.findOne({ awid: awid });
  }

  getByCode(awid, code) {
    return super.findOne({ awid: awid, code: code });
  }

  getByCodes(awid, codes) {
    return super.find(
      {
        awid: awid,
        code: {
          $in: codes
        }
      },
      {},
      {},
      {
        _id: 0,
        code: 1,
        roleUri: 1
      }
    );
  }

  getCount(awid) {
    return super.count({ awid });
  }

  update(uuObject) {
    uuObject = { ...uuObject };

    let awid = uuObject["awid"];
    delete uuObject["awid"];

    let code = uuObject["code"];
    delete uuObject["code"];

    return super.findOneAndUpdate(
      {
        code: code,
        awid: awid
      },
      uuObject,
      "NONE"
    );
  }

  deleteByAwid(awid) {
    return super.deleteMany({ awid: awid });
  }

  deleteByCode(awid, code) {
    return super.deleteOne({ awid, code });
  }

  listByCode(awid, codes = null, pageInfo = {}) {
    let filter = { awid };

    if (Array.isArray(codes) && codes.length) {
      filter.code = { $in: codes };
    } else if (typeof codes == "string") {
      filter.code = codes;
    }

    return super.find(filter, pageInfo, { "sys.cts": -1 });
  }
}

module.exports = SysProfileMongo;
