// uu_appg01_workspace/src/controllers/sys-app-workspace-controller.js
"use strict";

const SysAppWorkspaceModel = require("../models/sys-app-workspace-model.js");

class SysAppWorkspaceController {
  initAppWorkspace(ucEnv) {
    return SysAppWorkspaceModel.initAppWorkspace(ucEnv.parameters);
  }

  getAppWorkspace(ucEnv) {
    return SysAppWorkspaceModel.getAppWorkspace(ucEnv.uri, ucEnv.parameters);
  }

  getLicenseOwner(ucEnv) {
    return SysAppWorkspaceModel.getLicenseOwner(ucEnv.uri.getAwid());
  }

  setAuthorizationType(ucEnv) {
    return SysAppWorkspaceModel.setAuthorizationType(ucEnv.uri, ucEnv.parameters, ucEnv.session);
  }

  connectArtifact(ucEnv) {
    return SysAppWorkspaceModel.connectArtifact(ucEnv.uri, ucEnv.parameters, ucEnv.session);
  }

  setAppWorkspaceState(ucEnv) {
    return SysAppWorkspaceModel.setAppWorkspaceState(ucEnv.uri, ucEnv.parameters);
  }

  deleteAppWorkspace(ucEnv) {
    return SysAppWorkspaceModel.deleteAppWorkspace(ucEnv.uri, ucEnv.parameters);
  }

  setAwidRuntimeMode(ucEnv) {
    return SysAppWorkspaceModel.setAwidRuntimeMode(ucEnv.uri.getAwid(), ucEnv.parameters);
  }

  getAwidRuntimeMode(ucEnv) {
    return SysAppWorkspaceModel.getAwidRuntimeMode(ucEnv.uri.getAwid(), ucEnv.parameters);
  }
}

module.exports = new SysAppWorkspaceController();
