// uu_appg01_workspace/src/controllers/sys-dump-restore-controller.js
"use strict";

const model = require("../models/sys-dump-restore-model");

class SysDumpRestoreController {
  async dump(ucEnv) {
    const data = await model.dump(ucEnv.getDtoIn());
    const filename = `uuAppDataStoreDump_${new Date()
      .toISOString()
      .split("T")[0]
      .replace(/-/g, "_")}.zip`;

    return ucEnv.getResponse().setBinaryBody(data, { filename, contentType: "application/zip" });
  }

  async restore(ucEnv) {
    return await model.restore(ucEnv.uri, ucEnv.parameters);
  }

  async awidDump(ucEnv) {
    const data = await model.awidDump(ucEnv.uri.getAwid());
    const filename = `uuAppDataStoreDump_${new Date()
      .toISOString()
      .split("T")[0]
      .replace(/-/g, "_")}.zip`;

    return ucEnv.getResponse().setBinaryBody(data, { filename, contentType: "application/zip" });
  }

  async awidRestore(ucEnv) {
    return await model.awidRestore(ucEnv.uri, ucEnv.parameters);
  }

  import(ucEnv) {
    return model.import(ucEnv.getUri().getAwid(),ucEnv.getDtoIn());
  }

  deleteImportMappings(ucEnv) {
    return model.deleteImportMappings(ucEnv.getUri().getAwid());
  }

  listImportMappings(ucEnv) {
    return model.listImportMappings(ucEnv.getUri().getAwid(),ucEnv.getDtoIn());
  }

}

module.exports = new SysDumpRestoreController();
